File: entities/monorem.js
let _monoremsJointEnergy = 150;


class Monorem extends Attacker {
    constructor() {
        super();
    }

    _next_tick() {
        if (this.isNear(Air) && this.can_multiply()) {
            let cell = random(this.findNear(Air));
            this.multiply(cell);
        }

        else if (this.isNear(Xotaker, Predator, Creeper)) {
            let cell = random(this.findNear(Xotaker, Predator, Creeper));
            this.eat(cell);
        }

        else if (this.isNear(Air, Grass)) {
            let cell = random(this.findNear(Air, Grass));
            this.move(cell);
        }
    }

    get energy() {
        if ("_game_logic" in this) {
            return _monoremsJointEnergy/this._game_logic.count(this.kind);
        }
        return this._default_energy;
    }

    set energy(value) {
        _monoremsJointEnergy -= this.energy;
        _monoremsJointEnergy += value;
        if (this.energy > this._energy_limit) {
            _monoremsJointEnergy = this._energy_limit*this._game_logic.count(this.kind);
        }
    }
}

Monorem._default_energy = 5;
Monorem._multiplication_cost = 9;
Monorem._energy_limit = 20;
Monorem._energy_from_prey = 3;

File: entities/monorem.py
from random import choice

from entities.air import Air
from entities.grass import Grass
from entities.mob import Mob


class Monorem(Mob):
    __slots__ = ()

    _default_energy = 15
    energy_limit = 40
    _energy_increment_per_tick = 1

    def _next_tick(self):
        if self._do_multiply_on(Air) and self._is_near(Air):
            pos = choice(self._find_near(Air)).pos
            self._multiply(pos)

        elif self._do_multiply_on(Grass) and self._is_near(Grass):
            pos = choice(self._find_near(Grass)).pos
            self._kill(pos)
            self._multiply(pos)

        elif self._is_near(Air, Grass):
            pos = choice(self._find_near(Air, Grass)).pos
            self.field.swap(self.pos, pos)

    def _do_multiply_on(self, kind):
        if kind is Grass:
            return self.energy >= 2*self._default_energy + self._energy_increment_per_tick
        return super()._do_multiply_on(kind)
